src/sage/manifolds/differentiable/tensorfield.py

```diff
@@ -3031,17 +3031,28 @@ def __call__(self, *args):
             resu._latex_name = res_latex
             return resu
 
-    def trace(self, pos1=0, pos2=1):
+    def trace(
+        self,
+        pos1=0,
+        pos2=1,
+        using: Optional[
+            Union[PseudoRiemannianMetric, SymplecticForm, PoissonTensorField]
+        ] = None,
+    ):
         r"""
         Trace (contraction) on two slots of the tensor field.
 
+        If a non-degenerate form is provided, the trace of a `(0,2)` tensor field
+        is computed by first raising the last index.
+
         INPUT:
 
         - ``pos1`` -- (default: 0) position of the first index for the
           contraction, with the convention ``pos1=0`` for the first slot
         - ``pos2`` -- (default: 1) position of the second index for the
           contraction, with the same convention as for ``pos1``. The variance
           type of ``pos2`` must be opposite to that of ``pos1``
+        - ``using`` -- (default: ``None``) a non-degenerate form
```

Comment:
How about ``form``?

Reply:
There are already to many forms involved here, with `a.trace(form=b)` its not clear if you take the trace of a or of b. With `a.trace(using=b)` you don't have this issue.

Reply:
I am not sure about adding the new argument `using` (or whatever its name) to the function `trace` of generic tensor fields for just treating the specific case of type (0,2) tensor.  Is this really necessary? This adds some complexity to the interface and looks like a hack for dealing with a specific case, for which the concept of trace is not meaningfull a priori.

Reply:
I would argue that half of differential geometry concerns objects that eat two vector fields: symplectic forms, Riemannian metrics, curvature tensors etc. For example, the scalar curvature is the metric-trace of the Ricci tensor.

Reply:
Actually I am more against treating  the special case of a tensor of type (0,2) here (*) than adding a new argument to `trace`  for a new (shortcut) functionality. Why not making this available to all kinds of tensor fields ? i.e. if  `pos1` and `pos2` are of the same type
(both covariant or both contravariant), the trace is taken after `pos2` is raised or lowered by means of `using`. 

(*) this looks like bad code design: if tensors of type (0,2) are specific, they should deserve their own class

Reply:
Yes, this makes sense and is a really nice idea! Would it be okay though to implement this as a follow-up after https://github.com/sagemath/sage/pull/35448 (which would make it easier to detect if they indicies have the same type)?